packages/react-day-picker/src/contexts/Focus/FocusContext.tsx

```diff
@@ -59,35 +96,148 @@ export function FocusProvider({
   const focusDayBefore = () => {
     if (!focusedDay) return;
     const before = addDays(focusedDay, -1);
-    setDay(before);
+    changeInitialFocus(before);
     switchMonth(before, numberOfMonths * -1);
   };
   const focusDayAfter = () => {
     if (!focusedDay) return;
     const after = addDays(focusedDay, 1);
-    setDay(after);
+    changeInitialFocus(after);
     switchMonth(after, numberOfMonths);
   };
   const focusWeekBeforeDay = () => {
     if (!focusedDay) return;
     const up = addWeeks(focusedDay, -1);
-    setDay(up);
+    changeInitialFocus(up);
     switchMonth(up, numberOfMonths * -1);
   };
   const focusWeekAfterDay = () => {
     if (!focusedDay) return;
     const down = addWeeks(focusedDay, 1);
-    setDay(down);
+    changeInitialFocus(down);
     switchMonth(down, numberOfMonths);
   };
 
+  const focusFirstDayOfWeek = (): void => {
+    if (!focusedDay) return;
+    const dayToFocus = addDays(
+      startOfWeek(addDays(focusedDay, -1), {
+        weekStartsOn: 0
+      }),
+      1
+    );
+    goToMonth(dayToFocus);
+    changeInitialFocus(dayToFocus);
+  };
+
+  const focusLastDayOfWeek = (): void => {
+    if (!focusedDay) return;
+    const dayToFocus = addDays(
+      endOfWeek(addDays(focusedDay, -1), {
+        weekStartsOn: 0
+      }),
+      1
+    );
+    goToMonth(dayToFocus);
+    changeInitialFocus(dayToFocus);
+  };
+
+  interface FocusOptions {
+    unit: 'month' | 'year';
+    direction: 'backwards' | 'forward';
+  }
+
+  const focusByUnit = (opts: FocusOptions): void => {
+    if (!focusedDay) return;
+    if (!tempoFocusedDay) tempoFocusDay(focusedDay);
+
+    const { unit, direction } = opts;
+    let navigationFunction: (date: Date | number, amount: number) => Date;
+
+    switch (unit) {
+      case 'month':
+        switch (direction) {
+          case 'forward':
+            navigationFunction = addMonths;
+            break;
+          case 'backwards':
+            navigationFunction = subMonths;
+            break;
+        }
+        break;
+      case 'year':
+        switch (direction) {
+          case 'forward':
+            navigationFunction = addYears;
+            break;
+          case 'backwards':
+            navigationFunction = subYears;
+            break;
+        }
+    }
+
+    const month = navigationFunction(focusedDay, 1);
+    goToMonth(month);
+
+    initialFocusChanged.current = Boolean(
+      tempoFocusedDay && getDaysInMonth(month) < tempoFocusedDay.getDate()
+    );
+
+    focus(
+      !initialFocusChanged.current
+        ? tempoFocusedDay
+          ? addDays(startOfMonth(month), tempoFocusedDay.getDate() - 1)
+          : month
+        : addDays(
+            startOfWeek(addWeeks(startOfMonth(month), 3)),
+            tempoFocusedDay ? tempoFocusedDay.getDay() : focusedDay.getDay()
+          ),
+      true
+    );
+  };
+
+  const focusMonthBefore = () => {
+    focusByUnit({
+      direction: 'backwards',
+      unit: 'month'
+    });
+  };
```

Comment:
These and the other focus function should be written with the same code style as the others:

```suggestion
  const focusMonthBefore = () => {
    if (!focusedDay) return;
    const monthBefore = addMonths(focusedDay, -1);
    setDay(monthBefore);
    switchMonth(monthBefore, numberOfMonths);
   };
```

Reply:
Alright, will do.